enforcer/enforcer_main.py

```diff
@@ -0,0 +1,249 @@
+from pydantic import BaseModel
+from prometheus_client import generate_latest, CONTENT_TYPE_LATEST
+
+from enforcer.dal.supabase_dal import SupabaseDal
+from enforcer.patch_manager import patch_container_resources
+from enforcer.model import PodOwner, WorkloadRecommendation
+from enforcer.env_vars import ENFORCER_SSL_KEY_FILE, ENFORCER_SSL_CERT_FILE, KRR_MUTATION_MODE_DEFAULT
+from enforcer.resources.owner_store import OwnerStore
+from enforcer.resources.recommendation_store import RecommendationStore
+from enforcer.metrics import pod_admission_mutations, replicaset_admissions, rs_owners_size, admission_duration
+
+# Configure logging
+logger = logging.getLogger()
+logHandler = logging.StreamHandler(sys.stdout)
+formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')
+logHandler.setFormatter(formatter)
+logger.addHandler(logHandler)
+logger.setLevel(os.environ.get("LOG_LEVEL", "INFO"))
+
+# Define the mention pattern regex
+MENTION_PATTERN = re.compile(r'@[\w.-]+')
+ENFORCE = "enforce"
+IGNORE = "ignore"
+
+app = FastAPI(
+    title="KRR Enforcer mutation webhook",
+    description="A KRR recommendations mutating webhook server for Kubernetes",
+    version="1.0.0"
+)
+
+dal = SupabaseDal()
+recommendation_store = RecommendationStore(dal)
+owner_store = OwnerStore()
+
+class AdmissionReview(BaseModel):
+    apiVersion: str
+    kind: str
+    request: Dict[str, Any]
+
+def admission_allowed(request: AdmissionReview) -> Dict[str, Any]:
+    return \
+        {
+            "apiVersion": "admission.k8s.io/v1",
+            "kind": "AdmissionReview",
+            "response": {
+                "uid": request.request.get('uid'),
+                "allowed": True
+        }
+    }
+
+def enforce_pod(pod: Dict[str, Any]) -> bool:
+    mode = pod.get('metadata', {}).get('annotations', {}).get("admission.robusta.dev/krr-mutation-mode", None)
+    if mode == ENFORCE:
+        return True
+    elif mode == IGNORE:
+        return False
+    else:
+        return KRR_MUTATION_MODE_DEFAULT == ENFORCE
+
+
+@app.post("/mutate")
+async def mutate(request: AdmissionReview):
+    """
+    Handle mutating webhook requests from Kubernetes.
+    
+    Args:
+        request (AdmissionReview): The admission review request from Kubernetes
+        
+    Returns:
+        dict: Admission review response
+    """
+    start_time = time.time()
+    try:
+        logging.debug(f"Admission request received %s", request)
+        # Extract the object being reviewed
+        object_to_review = request.request.get('object', {})
+        kind = request.request.get('kind', {}).get('kind')
+
+        if kind == "ReplicaSet":  # use create/delete admission requests, to track new/removed replica sets owners
+            operation = request.request.get('operation', 'UNKNOWN')
+            replicaset_admissions.labels(operation=operation).inc()
+            admission_duration.labels(kind='ReplicaSet').observe(time.time() - start_time)
```

Comment:
should you update admission_duration lower down (i.e. after updating the owner store) so its more accurate?